File: cartpoleKalman.py
from datetime import datetime
from blitting import LivePlot
from cartpoleEnv import CartPoleEnv
from tinyekf import EKF
import numpy as np

env = CartPoleEnv(render_mode='human', noise_lvl=0.05)
state = env.reset()
action = 0

lp = LivePlot(
    labels=(
        ('x_obs', 'x_est'), 
        ('theta_obs', 'theta_est'),
        ('xdot_obs', 'xdot_est'), 
        ('thetadot_obs', 'thetadot_est'),
    ),
    ymins=[-2.4, -np.pi/4, -1, -1],
    ymaxes=[2.4, np.pi/4, 1, 1],
)

class PitchEKF(EKF):
    def __init__(self):
        # 4 states, 1 observable
        EKF.__init__(self, n=4, m=4, pval=0.1, qval=0.5, rval=0.25)

    def f(self, state, action):
        x, x_dot, theta, theta_dot = state
        force = action

        gravity = 9.8
        masscart = 1.0
        masspole = 0.15
        total_mass = masspole + masscart
        length = 0.5  # actually half the pole's length
        polemass_length = masspole * length
        dt = 1/30.

        costheta = np.cos(theta)
        sintheta = np.sin(theta)

        temp = (
            force + polemass_length * theta_dot**2 * sintheta
        ) / total_mass
        thetaacc = (gravity * sintheta - costheta * temp) / (
            length * (4.0 / 3.0 - masspole * costheta**2 / total_mass)
        )
        xacc = temp - polemass_length * thetaacc * costheta / total_mass

        x = x + dt * x_dot
        x_dot = x_dot + dt * xacc
        theta = theta + dt * theta_dot
        theta_dot = theta_dot + dt * thetaacc

        new_state = np.array([x, x_dot, theta, theta_dot])
        state_dot = np.array([x_dot, xacc, theta_dot, thetaacc])

        return new_state, state_dot

    def h(self, obs):
        #imu will measure pitch (theta), at the 2nd index
            
        # H = np.array([1, 1, 1, 1])
        H = np.eye(4)
        h = np.dot(H, obs)

        return h, H

pitchEKF = PitchEKF()

pitchEKF.x = state


while True:

    #x, xdot, theta, thetadot
    K = [0.5, 0.3, 20, 1]

    obs, reward, done, state = env.step(action)

    x_obs, xdot_obs, theta_obs, thetadot_obs = obs
    x, xdot, theta, thetadot = state

    state_est = pitchEKF.step(obs, action)

    x_est, xdot_est, theta_est, thetadot_est = state_est

    action = K @ np.array([x_est, xdot_est, theta_est, thetadot_est])

    lp.plot(
        x_obs, x_est, 
        theta_obs, theta_est,
        xdot_obs, xdot_est,
        thetadot_obs, thetadot_est,
    )

    env.render()
    if done == True:
        state = env.reset()
        pitchEKF.reset(state)
        action=0

File: cartpoleKalman.cpp
#include <iostream>
// #include <Eigen/Dense>
#include "Eigen/Dense"
#include <math.h>

using Eigen::Vector4f;
using Eigen::Matrix4f;
using Eigen::MatrixXd;

//4 states: [x, xdot, theta, thetadot]
Vector4f state, statedot;
Matrix4f P_pre, P_post;     //prediction covariance
Matrix4f Q;                 //process covariance
Matrix4f R;                 //measurement noise covariance
Vector4f obs;               //observations
Matrix4f H;                 //selects which states are observed
Vector4f h;                 //observed states

const float gravity = 9.8f;
const float masscart = 1.0f;
const float masspole = 0.1f;
const float total_mass = masspole + masscart;
const float length = 0.5f;
const float polemass_length = masspole * length;

Matrix4f I_4;

float dt = 1/30.0f;
float force = 0.0f;

//model prediction (f)
void predict(){
    float x = state(0);
    float xdot = state(1);
    float theta = state(2);
    float thetadot = state(3);

    float sintheta = sin(theta);
    float costheta = cos(theta);

    float temp = (force + polemass_length * thetadot*thetadot * sintheta) / total_mass;
    float thetaacc = (gravity * sintheta - costheta * temp) / (length * (4.0 / 3.0 - masspole * costheta*costheta / total_mass));
    float xacc = temp - polemass_length * thetaacc * costheta / total_mass;

    //Euler integration
    x += dt * xdot;
    xdot += dt * xacc;
    theta += dt * thetadot;
    thetadot += dt * thetaacc;

    state(0) = x;
    state(1) = xdot;
    state(2) = theta;
    state(3) = thetadot;

    statedot(0) = xdot;
    statedot(1) = xacc;
    statedot(2) = thetadot;
    statedot(3) = thetaacc;
}

//select which states are being observed
void form_obs(){
    H = I_4; //observe all 4 states
    h = H * obs;
}

void ekf_init(float pval, float qval, float rval){
    I_4 <<  1, 0, 0, 0,
            0, 1, 0, 0,
            0, 0, 1, 0,
            0, 0, 0, 1;

    state << 0, 0, 0, 0;
    P_post = I_4 * pval;
    Q = I_4 * qval;
    R = I_4 * rval;
}

//one step of the EKF using the observation obs
void step(){
    predict(); //updates state and statedot using force and dt
    P_pre = statedot * P_post * statedot + Q;

    form_obs(); //updates h and H, the formatted observations
    Matrix4f H_t = H.transpose();

    Matrix4f G = (P_pre*H_t) * ((H*P_pre*H_t).inverse() + R);

    state = state + G*(obs - h);



}


int main(){


    MatrixXd m(2, 2);
    m(0, 0) = 3;
    m(1, 0) = 3.42;
    m(0, 1) = -1;
    m(1, 1) = m(1, 0) + m(0, 1);
    std::cout << m << std::endl;
}